### ISSUE TITLE
Automerge label

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#11436]
### Problem

Often people review a PR and approve it, but the CI tests take an hour or more to complete, so it isn't merged when it is approved. People have to remember to come back and merge a PR that passed all the tests.

### Proposed Solution

In conda-forge, there is an `automerge` label, which if set, will cause the PR to be merged as soon as tests pass successfully. We could introduce something like that so that a reviewer could give a positive review and add the automerge label, and not have to come back to the PR to merge it once tests pass successfully.

Thinking conservatively, I would imagine we would want automerge to merge only if the PR was pretty uncontroversial:

1. The PR had at least one positive review and only positive reviews.
2. The PR passes tests.
3. The PR has the automerge label applied.
4. The PR is not draft.
5. The PR has been open at least some amount of time to make sure people have a chance to see it (maybe 48 hours?)

**Comment by jasongrout** [comment]
It looks like Github natively supports some sort of automerge functionality: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request

**Comment by jtpio** [comment]
For reference this is where the check is triggered:

https://github.com/jupyterlab/jupyterlab/blob/a7443fa660897f69cadd25c27f86df86f114fb2f/.github/workflows/benchmark.yml#L11

Looks like it is triggered on "PR approved" (or on demand via a PR comment) to not run that check on every change since it takes a while to complete.

> It looks like Github natively supports some sort of automerge functionality: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request

Let's enable it and see how it goes?

In the settings we would have to check that box:

![image](https://user-images.githubusercontent.com/591645/141769406-a721bc48-ee3a-4b4b-aa5d-6009ba9c8dd9.png)

**Comment by krassowski** [comment]
I will just highlight that it also requires setting `Require pull request reviews before merging` or `Require status checks to pass before merging` in branches protection rules - this was not very clear for me when I was setting it up earlier on the LSP repo. You can set just one CI job as required and it will suffice to enable enabling auto-merge.

**Comment by jtpio** [comment]
To answer a question raised during the [weekly call today](https://github.com/jupyterlab/team-compass/issues/128#issuecomment-971829367):

> After you enable auto-merge for a pull request, if someone who does not have write permissions to the repository pushes new changes to the head branch or switches the base branch of the pull request, auto-merge will be disabled. For example, if a maintainer enables auto-merge for a pull request from a fork, auto-merge will be disabled after a contributor pushes new changes to the pull request.

**Comment by jtpio** [comment]
@fcollonval I'm not sure it's possible to wait a certain amount of time before auto merging.

Maybe it's possible to work around it by having an extra CI check that would implement some kind of waiting logic.

**Comment by jasongrout** [comment]
> Maybe it's possible to work around it by having an extra CI check that would implement some kind of waiting logic.

Oh, interesting: a CI check that takes a day or two to return success?

**Comment by krassowski** [comment]
But there are also concurrent jobs limits and minutes limits. While it sounds interesting, I don't think that we should do it. Also, I don't see a good motivation for having the extra waiting period if using the native GitHub automerge - only commiters can trigger it and the fact of enabling automerge is shown on the PR timeline (along with who triggered it) - it is more of a "I would merge it now if the CI was green" rather than "I will trust the bot to decide when to merge it" (conda-forge-style).

Of course if going for fully-automated conda-forge (or old dependabot)-style automerge which would kick in automatically (/upon labelling by anyone with triage privilages) that time delay condition would be necessary.

**Comment by krassowski** [comment]
Clarifying: I support the motion for commiters to wait for at least 24/48 hours before merging anything which is not trivial. Just saying that adding a new CI job to coerce GithHub-native automerge to work like that may not be beneficial. I think that the former could be solved by having some guidelines on merging (like what is our etiquette on merging and reviews? who can merge? when to merge? if a non-commiter sees a a green button on a backport because a branches is not protected, should they press it?)

**Comment by jasongrout** [comment]
I agree that tying up our concurrent github actions for a time delay is not very feasible. I was imagining a CI service that would just respond in a 24-48 hours, but not actually do any work. Or maybe we just say that the 1-2 day wait is not on the last commit, but only from the PR creation date. The idea is to give people a chance to see it if it isn't trivial.

**Comment by jasongrout** [comment]
I haven't written a GitHub CI check, but I imagine something like this can work:

0. Decide a threshold for a PR to be open. Say it is N days.
1. register a webhook for check suite creations (which I imagine are what happens when a CI run is started?)
2. If the PR has been open at least N days, immediately [post](https://docs.github.com/en/rest/reference/repos#create-a-commit-status) a `success` commit status
3. If the PR has been open less than N days, immediately post a `pending` commit status. Schedule a timer (or just have a batch job that runs each day) to post a `success` commit status when the PR is N days old.

Surely there is some sort of automated something out there to make this sort of workflow trivial to implement?

**Comment by fcollonval** [comment]
From the previous comments, the easiest solution seems to create a GitHub action triggered periodically (like once a day) that will go through all opened non-draft PR, merge it when a set of conditions is met and add a comment on the PR.

The conditions:
- Merge policy:
  - CI passes
  - PR approved
  - Last activity > _N_ days
- Technical constrains:
  - Must contain a label required for automatic changelog generation
  - Must target a milestone
  - No merge conflict

> We will need to be careful adding some delay between merge triggers (see https://octokit.github.io/rest.js/v18#pulls-merge and [secondary rate limits](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits)) - especially because one merge can create conflict for others.
> We need to look at stale PRs to ensure they won't match the merge criteria.

**Comment by jasongrout** [comment]
I would prefer a slightly different set of conditions for automerging.

1. CI passes
2. PR approved by someone with "commit rights", and not disapproved by anyone. Note that it may also be approved by a number of other people, but for automerging we should make sure at least one of the approvers has "commit rights"
3. The PR has been open for at least N days
4. Must contain a label for automatic changelog generation
5. Must target a milestone
6. No merge conflict

This enables us to move more quickly (we just need to make sure people have a chance to see the PR being opened, rather than waiting for there to be no more discussion or activity of any kind), and it lets everyone review PRs without those reviews always triggering a merge.

**Comment by jasongrout** [comment]
Another way we can have the built-in automerge works is perhaps we say that meeseeksdev (or jupyterlab-bot?) must approve automerge PRs, and the bot only approves when:

1. The PR has been open for at least N days
2. Must contain a label for automatic changelog generation
3. Must target a milestone

**Comment by jasongrout** [comment]
> I would prefer a slightly different set of conditions for automerging.

Even this doesn't allow for a committer to approve, but ask someone else to also look at it. I think I would add a step that is a manual request from a person with commit rights to automerge the PR. That's one thing I really like about using the built-in automerging - it is an explicit request to automerge a PR, independent of any reviewing or other activity.
